mem/buffer_pool.go

```diff
@@ -0,0 +1,156 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package mem
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+)
+
+// BufferPool is a pool of buffers that can be shared, resulting in
+// decreased memory allocation. Currently, in gRPC-go, it is only utilized
+// for parsing incoming messages.
+//
+// # Experimental
+//
+// Notice: This API is EXPERIMENTAL and may be changed or removed in a
+// later release.
+type BufferPool interface {
+	// Get returns a buffer with specified length from the pool.
+	//
+	// The returned byte slice may be not zero initialized.
+	Get(length int) []byte
+
+	// Put returns a buffer to the pool.
+	Put([]byte)
+}
+
+var defaultBufferPoolSizes = []int{
+	256,
+	4 << 10,  // 4KB
+	16 << 10, // 16KB (max HTTP/2 frame size used by gRPC)
+	64 << 10, // 64KB
+	1 << 20,  // 1MB
+}
+
+var defaultBufferPool = func() *atomic.Pointer[BufferPool] {
+	pool := NewBufferPool(defaultBufferPoolSizes...)
+	ptr := new(atomic.Pointer[BufferPool])
+	ptr.Store(&pool)
+	return ptr
+}()
+
+func DefaultBufferPool() BufferPool {
+	return *defaultBufferPool.Load()
+}
+
+func SetDefaultBufferPool(pool BufferPool) {
+	defaultBufferPool.Store(&pool)
+}
+
+func NewBufferPool(poolSizes ...int) BufferPool {
+	sort.Ints(poolSizes)
+	pools := make([]*sizedBufferPool, len(poolSizes))
+	for i, s := range poolSizes {
+		pools[i] = newBufferPool(s)
+	}
+	return &tieredBufferPool{
+		sizedPools:   pools,
+		fallbackPool: newBufferPool(0),
+	}
+}
+
+// tieredBufferPool implements the BufferPool interface with multiple tiers of
+// buffer pools for different sizes of buffers.
+type tieredBufferPool struct {
+	sizedPools   []*sizedBufferPool
+	fallbackPool *sizedBufferPool
```

Comment:
I'm not completely convinced with using a `sizedBufferPool` for fallback. Could we instead use something like a `nopBufferPool` from the now deleted `shared_buffer_pool.go`. The reason I'm saying this is because:
- a fallback pool created as a `sizedBufferPool` of size `0`, will always return an empty slice from its `New` function, and `Get` will always add this back to the pool because `if cap(bs) < size` will always be `true`
- and if we don't expect the `sizedBufferPool` to work with a size of `0`, we can get rid of the extra code in `Get`, which can now simply be:
```golang
// Get assumes that size is always <= defaultSize
func (p *sizedBufferPool) Get(size int) []byte {
	bs := *p.pool.Get().(*[]byte)
	return bs[:size]
}
```

Reply:
You are right, there's a very good chance the current implementation of `Get` will initially dump a bunch of empty buffers in the pool. The old `nopBufferPool` is actually preserved here, but it's not clever enough IMO because it doesn't actually pool buffers, it just creates them on demand. I will add a new `simpleBufferPool` implementation that does on-demand stuff but still pools the buffers as best it can.

Reply:
Done